librepo/handle.c

```diff
@@ -841,7 +867,57 @@ lr_handle_setopt(LrHandle *handle,
     return ret;
 }
 
-static gboolean
+gboolean
+lr_handle_network_wait(LrHandle *handle, GError **err, guint seconds)
+{
+
+    assert(!err || *err == NULL);
+
+    if (!handle) {
+        g_set_error(err, LR_HANDLE_ERROR, LRE_BADFUNCARG,
+                    "No handle specified");
+        return FALSE;
+    }
+
+    GNetworkMonitor *monitor = g_network_monitor_get_default();
+    GCancellable *cancellable = g_cancellable_new ();
+
+    struct callback_data data_struct;
+    data_struct.input_seconds = seconds;
+    data_struct.begin = time(NULL);
+    data_struct.cancellable = cancellable;
+    data_struct.monitor = monitor;
+
+    gchar *baseurl;
+    if(handle->metalinkurl)
+      baseurl = handle->metalinkurl;
+    else if(handle->mirrorlisturl)
+      baseurl = handle->mirrorlisturl;
+    else if(handle->urls)
+      baseurl = handle->urls[0];
+    assert(baseurl);
+
+    GUri *uri = g_uri_parse(baseurl, G_URI_FLAGS_NONE, NULL);
+    if(uri == NULL){
+      return FALSE;
+    }
+    const gchar* host = g_uri_get_host(uri);
+    guint16 port = g_uri_get_port(uri);
+    GSocketConnectable *connectable =  g_network_address_new(host, port);
+    data_struct.connectable = connectable;
+    GMainLoop *loop;
+    loop = g_main_loop_new(NULL, FALSE);
+    data_struct.loop = loop;
+    g_timeout_add(200, timeout_callback, &data_struct); 
+    g_main_loop_run(loop);
+    g_main_loop_unref(loop);
```

Comment:
We can use `g_autoptr(GMainLoop) loop = g_main_loop_new(NULL, FALSE);` above, then drop this manual unref.